lib/ansible/modules/windows/win_domain_user.ps1

```diff
@@ -4,6 +4,109 @@
 
 #Requires -Module Ansible.ModuleUtils.Legacy
 
+Function Test-Credential {
+    param(
+        [String]$Username,
+        [String]$Password,
+        [String]$Domain = $null,
+        $LOGON32_PROVIDER = 0, # Default
+        $LOGON32_LOGON_TYPE = 3 # NetworkLogon
+    )
+    if (($Username.ToCharArray()) -contains [char]'@') {
+        # UserPrincipalName
+        $Username = ($Username -split '@')[0]
+        $Domain = ($Username -split '@')[1]
+    } elseif (($Username.ToCharArray()) -contains [char]'\') {
+        # Pre Win2k Account Name
+        $Username = ($Username -split '\')[0]
+        $Domain = ($Username -split '\')[1]
+    } else {
+        # No domain provided, so maybe local user?
+    }
+    $null = $Domain # Make CI-Check happy...
+
+    # More information about LogonUser at https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-logonusera
+    $platform_util = @'
+using System;
+using System.Runtime.InteropServices;
+using System.ComponentModel;
+using System.Security;
+namespace Ansible
+{
+    public class WinUserPInvoke
+    {
+        const int LOGON32_LOGON_NETWORK           = 3;
+        const int LOGON32_PROVIDER_DEFAULT        = 0; // Negotiate if domain == null and username not upn else NTLM
+        const int LOGON32_PROVIDER_WINNT40        = 2; // Negotiate
+        const int LOGON32_PROVIDER_WINNT50        = 3; // NTLM
+        // Refere to: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d
+        static int[] success_codes = new int[] {
+            0x0000052F,  // ERROR_ACCOUNT_RESTRICTION
+            0x00000530,  // ERROR_INVALID_LOGON_HOURS
+            0x00000531,  // ERROR_INVALID_WORKSTATION
+            0x00000569   // ERROR_LOGON_TYPE_GRANTED
+        };
+        static int[] failed_codes = new int[] {
+            0x0000052E,  // ERROR_LOGON_FAILURE - the user or pass was incorrect
+            0x00000006,  // ERROR_INVALID_HANDLE - the handle is invalid
+            0x00000532   // ERROR_PASSWORD_EXPIRED - The password has exired
+        };
+        [DllImport("advapi32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        internal static extern bool LogonUser(String username, String domain, IntPtr password,
```

Comment:
Handling the SecureString -> Ptr is really unecessary considering the original source is a string. It will be a lot less complicated if you just pass the password as a string and not a pointer.

Reply:
This is the recommended implementation microsoft in there examples like here: https://docs.microsoft.com/de-de/dotnet/api/system.runtime.interopservices.marshal.securestringtoglobalallocunicode?view=netframework-4.8

The password will spread around memory without being cleaned up. And I don't know if even the garbage collector will pick up copies to/in unmanaged code... Right now I've my doubts that this would be secure, I don't want to accidentally leak secrets just because security was inconvenient. Can you assure me that using a string here is as secure as my current approach?

Reply:
It is the recommended implementation if the input value is a secure string, in this case the input value is a string (passing in as a string from the controller). You are manually converting to a secure string then doing the whole copying to unmanaged memory then clearing it out there later. If the original value was a SecureString and we never dealt with just a String I would agree with you here but ultimately that is not the case.

> The password will spread around memory without being cleaned up

If you let the PInvoke marshaler deal with this you don't have to worry about it at all. Having the function definition be a string will make the .NET CLR automatically convert a string to unmanaged memory and pass in a pointer itself. This is also guaranteed by the CLR to be released once the function is returned. You are just skipping that step and doing it all yourself manually.

> Right now I've my doubts that this would be secure, I don't want to accidentally leak secrets just because security was inconvenient

Right now the password is already in memory as plaintext due to the input value from the controller just being a json string that is converted. There's nothing you can do about this as Ansible has no way of copying a secure string without it being a string originally.

> Can you assure me that using a string here is as secure as my current approach?

The marshaler does this all for you and guarantees the memory is cleared once the function has returned. Even if something goes horribly wrong the process will exit and also clear out the data in memory.

In the end you are trying to make something secure that is not already secure. This can't be done with how it works right now. If you can't trust the machine you are connecting to with Ansible then how can you trust it when you've promoted it as a domain controller.

Reply:
> The marshaler does this all for you and guarantees the memory is cleared once the function has returned. Even if something goes horribly wrong the process will exit and also clear out the data in memory.

I did not know that, that makes it way less important to care about.

> There's nothing you can do about this as Ansible has no way of copying a secure string without it being a string originally.

Until now, but there is another Ticket to address this issue (Ok the other issue is about non pipelining connectors writing the file with the secrets to disk, but it's basically the same problem).

> In the end you are trying to make something secure that is not already secure.

I just did not want to contribute to the problem, and because I did not know about the "automated cleanup" it looked a bit more important to think about ;-)

> If you can't trust the machine you are connecting to with Ansible then how can you trust it when you've promoted it as a domain controller.

It's not about trusting the machine now, but more of trusting the machine in the future, bootstrapping the environment before handing over to the customer, but you're right, it does not quite fit in here.

> and it's just making this module even harder to maintain.

I still disagree on that one, I consider both solutions equally difficult to maintain and both have there unique pitfalls and weaknesses. But anyway, I'm switching this for a string, as you're right, it is less important than I thought it was.